### ISSUE TITLE
test alexnet  accuracy is only 0.5

### ISSUE THREAD

**Issue by qiyuqing** [facebookarchive/caffe2#1087]
when i run alexnet ,the test accuracy is only 0.5 ,
but in caffe ,it can run 0.57. 
so i don't know  what is wrong in my code 

```
`def create_Alexnet(model,
    data,
    num_input_channels,
    num_labels,
    label = None):
    conv1 = brew.conv(
        model,
        data,
        "conv1",
        num_input_channels,
        96,
        11,
        ("XavierFill", {}),
        ("ConstantFill", {}),
        stride=4,
        pad=0
    )
    relu1 = brew.relu(model,conv1, "conv1")
    norm1 = brew.lrn(model,relu1, "norm1", size=5, alpha=0.0001, beta=0.75)
    pool1 = brew.max_pool(model,norm1, "pool1", kernel=3, stride=2)
    conv2 = brew.conv(
        model,
        pool1,
        "conv2",
        96,
        256,
        5,
        ("XavierFill", {}),
        ("ConstantFill", {"value": 0.1}),
        group=2,
        stride=1,
        pad=2
    )
    relu2 = brew.relu(model,conv2, "conv2")
    norm2 = brew.lrn(model,relu2, "norm2", size=5, alpha=0.0001, beta=0.75)
    pool2 = brew.max_pool(model,norm2, "pool2", kernel=3, stride=2)
    conv3 = brew.conv(
        model,
        pool2,
        "conv3",
        256,
        384,
        3,
         ("XavierFill", {'std': 0.01}),
        ("ConstantFill", {}),
        pad=1
    )
    relu3 = brew.relu(model,conv3, "conv3")
    conv4 = brew.conv(
        model,
        relu3,
        "conv4",
        384,
        384,
        3,
        ("XavierFill", {}),
        ("ConstantFill", {"value": 0.1}),
        group=2,
        pad=1
    )
    relu4 = brew.relu(model,conv4, "conv4")
    conv5 = brew.conv(
        model,
        relu4,
        "conv5",
        384,
        256,
        3,
        ("XavierFill", {}),
        ("ConstantFill", {"value": 0.1}),
        group=2,
        pad=1
    )
    relu5 = brew.relu(model,conv5, "conv5")
    pool5 = brew.max_pool(model,relu5, "pool5", kernel=3, stride=2)
    fc6 = brew.fc( model,
        pool5, "fc6", 256 * 6 * 6, 4096, ("XavierFill", {}),
        ("ConstantFill", {"value": 0.1})
    )
    relu6 = brew.relu(model,fc6, "fc6")
    dropout1 = brew.dropout(model,relu6, 'dropout1', ratio=0.5, is_test=0)

    fc7 = brew.fc(model,
        dropout1, "fc7", 4096, 4096, ("XavierFill", {}), ("ConstantFill", {"value": 0.1})
    )
    relu7 = brew.relu(model,fc7, "fc7")
    dropout2 = brew.dropout(model,relu7, 'dropout2', ratio=0.5, is_test=0)
    fc8 = brew.fc(model,
        dropout2, "fc8", 4096, 1000, ("XavierFill", {}), ("ConstantFill", {"value": 0.0})
    )
    (softmax, loss) = model.SoftmaxWithLoss(
            [fc8, label],
            ["softmax", "loss"],
     )
    return (softmax, loss)`
```
![tttt](https://user-images.githubusercontent.com/26154129/29270588-25666bd8-812a-11e7-895b-32b6729be9a7.jpg)

>

**Comment by harouwu** [comment]
What is your learning rate and optimizer for learning this AlexNet?

**Comment by qiyuqing** [comment]
@harouwu 
`logging.basicConfig()
log = logging.getLogger("resnet50_trainer")
log.setLevel(logging.DEBUG)

dyndep.InitOpsLibrary('@/caffe2/caffe2/distributed:file_store_handler_ops')
dyndep.InitOpsLibrary('@/caffe2/caffe2/distributed:redis_store_handler_ops')

g_iterCount = 0

def AddImageInput(model, reader, batch_size, img_size):
    '''
    Image input operator that loads data from reader and
    applies certain transformations to the images.
    '''
    data, label = brew.image_input(
        model,
        reader, ["data", "label"],
        batch_size=batch_size,
        use_caffe_datum=True,
        mean=128.,
        std=128.,
        scale=256,
        crop=img_size,
        mirror=1
    )

    data = model.StopGradient(data, data)


def SaveModel(args, train_model, epoch):
    prefix = "[]_{}".format(train_model._device_prefix, train_model._devices[0])
    predictor_export_meta = pred_exp.PredictorExportMeta(
        predict_net=train_model.net.Proto(),
        parameters=data_parallel_model.GetCheckpointParams(train_model),
        inputs=[prefix + "/data"],
        outputs=[prefix + "/softmax"],
        shapes={
            prefix + "/softmax": (1, args.num_labels),
            prefix + "/data": (args.num_channels, args.image_size, args.image_size)
        }
    )

    # save the train_model for the current epoch
    model_path = "%s/%s_%d.mdl" % (
        args.file_store_path,
        args.save_model_name,
        epoch,
    )

    # set db_type to be "minidb" instead of "log_file_db", which breaks
    # the serialization in save_to_db. Need to switch back to log_file_db
    # after migration
    pred_exp.save_to_db(
        db_type="minidb",
        db_destination=model_path,
        predictor_export_meta=predictor_export_meta,
    )


def RunEpoch(
    args,
    epoch,
    train_model,
    test_model,
    total_batch_size,
    num_shards,
    expname,
    explog,
):
    '''
    Run one epoch of the trainer.
    TODO: add checkpointing here.
    '''
    # TODO: add loading from checkpoint
    log.info("Starting epoch {}/{}".format(epoch, args.num_epochs))
    epoch_iters = int(args.epoch_size / total_batch_size / num_shards)
    for i in range(epoch_iters):
        global g_iterCount
        g_iterCount = g_iterCount+1
        # This timeout is required (temporarily) since CUDA-NCCL
        # operators might deadlock when synchronizing between GPUs.
        timeout = 600.0 if i == 0 else 60.0
        with timeout_guard.CompleteInTimeOrDie(timeout):
            t1 = time.time()
            workspace.RunNet(train_model.net.Proto().name)
            t2 = time.time()
            dt = t2 - t1

        fmt = "Finished iteration {}/{} of epoch {} ({:.2f} images/sec)"
        if(g_iterCount % 100 == 0):
          log.info(fmt.format(i + 1, epoch_iters, epoch, total_batch_size / dt))
          prefix = "{}_{}".format(
            train_model._device_prefix,
            train_model._devices[0])
          accuracy  = workspace.FetchBlob(prefix + '/accuracy')
          loss      = workspace.FetchBlob(prefix + '/loss')
          train_fmt = "Training loss: {}, accuracy: {}"
          log.info(train_fmt.format(loss, accuracy))

        if(g_iterCount % 1000 == 0):
           test_accuracy = 0
           if (test_model is not None):
              # Run 100 iters of testing
              ntests = 0
              test_accuracies = []
              for _ in range(0, 200):
                  workspace.RunNet(test_model.net.Proto().name)
                  for g in test_model._devices:
                      test_accuracy += np.asscalar(workspace.FetchBlob(
                          "{}_{}".format(test_model._device_prefix, g) + '/accuracy'
                      ))
                      ntests += 1
              test_accuracy /= ntests
              test_fmt = "test iter: {}, accuracy: {}"
              log.info(test_fmt.format(g_iterCount, test_accuracy))
    # TODO: add checkpointing
    return epoch + 1


def Train(args):
    # Either use specified device list or generate one
    if args.gpus is not None:
        gpus = [int(x) for x in args.gpus.split(',')]
        num_gpus = len(gpus)
    else:
        gpus = list(range(args.num_gpus))
        num_gpus = args.num_gpus

    log.info("Running on GPUs: {}".format(gpus))

    # Verify valid batch size
    total_batch_size = args.batch_size
    batch_per_device = total_batch_size // num_gpus
    assert \
        total_batch_size % num_gpus == 0, \
        "Number of GPUs must divide batch size"

    # Round down epoch size to closest multiple of batch size across machines
    global_batch_size = total_batch_size * args.num_shards
    epoch_iters = int(args.epoch_size / global_batch_size)
    args.epoch_size = epoch_iters * global_batch_size
    log.info("Using epoch size: {}".format(args.epoch_size))

    # Create ModelHelper object
    train_arg_scope = {
        'order': 'NCHW',
        'use_cudnn': True,
        'cudnn_exhaustice_search': True,
        'ws_nbytes_limit': (args.cudnn_workspace_limit_mb * 1024 * 1024),
    }
    train_model = model_helper.ModelHelper(
        name="resnet50", arg_scope=train_arg_scope
    )

    num_shards = args.num_shards
    shard_id   = args.shard_id
    rendezvous = None

      # Model building functions
    def create_resnet50_model_ops_alexnet(model, loss_scale):
        [softmax, loss] = resnet.create_Alexnet(
            model,
            "data",
            num_input_channels=args.num_channels,
            num_labels=args.num_labels,
            label="label",
        )
        #loss = model.AveragedLoss([xent], ["loss"])
        #loss = model.Scale(loss, scale=loss_scale)
        #model.Accuracy([softmax, "label"], "accuracy")
        return [loss]


    def add_optimizer(model):
        stepsz = 100000
        #stepsz = int(30 * args.epoch_size / total_batch_size / num_shards)
        optimizer.add_weight_decay(model, args.weight_decay)
        optimizer.build_sgd(
            model,
            args.base_learning_rate,
            momentum=0.9,
            nesterov=0,
            policy="step",
            stepsize=stepsz,
            gamma=0.1
        )

    # Input. Note that the reader must be shared with all GPUS.
    reader = train_model.CreateDB(
        "reader",
        db=args.train_data,
        db_type=args.db_type,
        num_shards=num_shards,
        shard_id=shard_id,
    )

    def add_image_input(model):
        AddImageInput(
            model,
            reader,
            batch_size=batch_per_device,
            img_size=args.image_size,
        )

    # Create parallelized model
    data_parallel_model.Parallelize(
        train_model,
        input_builder_fun       =add_image_input,
        forward_pass_builder_fun=create_resnet50_model_ops_alexnet,
        optimizer_builder_fun   =add_optimizer,
        devices                 =gpus,
        rendezvous              =rendezvous,
        optimize_gradient_memory=True,
        cpu_device=args.use_cpu,
    )

    # Add test model, if specified
    test_model = None
    if (args.test_data is not None):
        log.info("----- Create test net ----")
        test_arg_scope = {
            'order': "NCHW",
            'use_cudnn': True,
            'cudnn_exhaustive_search': True,
        }
        test_model = model_helper.ModelHelper(
            name="resnet50_test", arg_scope=test_arg_scope
        )

        test_reader = test_model.CreateDB(
            "test_reader",
            db=args.test_data,
            db_type=args.db_type,
        )

        def test_input_fn(model):
            AddImageInput(
                model,
                test_reader,
                batch_size=batch_per_device,
                img_size=args.image_size,
            )

        data_parallel_model.Parallelize(
            test_model,
            input_builder_fun=test_input_fn,
            forward_pass_builder_fun=create_resnet50_model_ops_alexnet,
            param_update_builder_fun=None,
            devices=gpus,
            cpu_device=args.use_cpu,
        )
        workspace.RunNetOnce(test_model.param_init_net)
        workspace.CreateNet(test_model.net)

    workspace.RunNetOnce(train_model.param_init_net)
    workspace.CreateNet(train_model.net)

    epoch = 0
    expname = "resnet50_gpu%d_b%d_L%d_lr%.2f_v2" % (
        args.num_gpus,
        total_batch_size,
        args.num_labels,
        args.base_learning_rate,
    )
    explog = experiment_util.ModelTrainerLog(expname, args)
    # Run the training one epoch a time
    while epoch < args.num_epochs:
        epoch = RunEpoch(
            args,
            epoch,
            train_model,
            test_model,
            total_batch_size,
            num_shards,
            expname,
            explog
        )

        # Save the model for each epoch
        SaveModel(args, train_model, epoch)

        model_path = "%s/%s_" % (
            args.file_store_path,
            args.save_model_name
        )
        # remove the saved model from the previous epoch if it exists
        if os.path.isfile(model_path + str(epoch - 1) + ".mdl"):
            os.remove(model_path + str(epoch - 1) + ".mdl")
`

i use resnet50 code ,
base_learning_rate = 0.01  and weight_decay = 0.0005

**Comment by qiyuqing** [comment]
it is dropout function , in train model   dropout1 = brew.dropout(model,relu6, 'dropout1', ratio=0.5, is_test=0)

in test model 
 dropout1 = brew.dropout(model,relu6, 'dropout1', ratio=0.5, is_test=1)

now alexnet accuracy is 0.57
INFO:resnet50_trainer:test iter: 634000, accuracy: 0.574331054688
